[AArch64][GlobalISel] Disable fixed-point iteration in all Combiners

tobias-stadler opened the pull request:
Disable fixed-point iteration in all AArch64 Combiners after #102163.
See inline comments for justification of test changes.

tobias-stadler commented:
CTMark O0:
```
Program                                       compile_instructions                        size..text
                                              base-O0              patch-O0        diff   base-O0    patch-O0   diff
7zip/7zip-benchmark                           141676688007.00      141409424579.00 -0.19% 1004756.00 1004756.00 0.00%
Bullet/bullet                                  61212623785.00       61017628199.00 -0.32%  873356.00  873356.00 0.00%
SPASS/SPASS                                    14400540825.00       14348447793.00 -0.36%  645136.00  645136.00 0.00%
tramp3d-v4/tramp3d-v4                          18450535356.00       18380583241.00 -0.38%  894656.00  894656.00 0.00%
kimwitu++/kc                                   22538775759.00       22426150589.00 -0.50%  813492.00  813492.00 0.00%
ClamAV/clamscan                                14536543546.00       14378298132.00 -1.09%  702832.00  702832.00 0.00%
lencod/lencod                                  12679661465.00       12525031047.00 -1.22%  820564.00  820564.00 0.00%
mafft/pairlocalalign                            6940116013.00        6851372697.00 -1.28%  423000.00  423000.00 0.00%
consumer-typeset/consumer-typeset              12193354451.00       12003246725.00 -1.56%  908844.00  908844.00 0.00%
sqlite3/sqlite3                                 4596588654.00        4512822506.00 -1.82%  459904.00  459904.00 0.00%
                           Geomean difference                                      -0.87%                       0.00%

```

CTMark O2:
```
Program                                       compile_instructions                        size..text
                                              base-O2              patch-O2        diff   base-O2    patch-O2  diff
7zip/7zip-benchmark                           207008981549.00      205988150072.00 -0.49% 824088.00  824088.00 0.00%
Bullet/bullet                                  99712646588.00       99179976524.00 -0.53% 515864.00  515864.00 0.00%
SPASS/SPASS                                    43192888774.00       42858161952.00 -0.77% 443056.00  443056.00 0.00%
kimwitu++/kc                                   41563609655.00       41217223921.00 -0.83% 461284.00  461284.00 0.00%
tramp3d-v4/tramp3d-v4                          65778761906.00       65208111169.00 -0.87% 570560.00  570560.00 0.00%
lencod/lencod                                  57111009799.00       56497616800.00 -1.07% 541960.00  541960.00 0.00%
ClamAV/clamscan                                53935475581.00       53350116553.00 -1.09% 456352.00  456352.00 0.00%
mafft/pairlocalalign                           32799063008.00       32439200934.00 -1.10% 320888.00  320888.00 0.00%
consumer-typeset/consumer-typeset              35357798970.00       34865643450.00 -1.39% 419436.00  419436.00 0.00%
sqlite3/sqlite3                                35219936375.00       34678565893.00 -1.54% 436160.00  436160.00 0.00%
                           Geomean difference                                      -0.97%                      0.00%

```

tobias-stadler committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.6-bogner-wip

[skip ci]

tobias-stadler committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.6-bogner-wip

tobias-stadler commented on llvm/test/CodeGen/AArch64/GlobalISel/combine-udiv.ll:
```diff
@@ -257,10 +257,10 @@ define i32 @udiv_div_by_180(i32 %x)
-; GISEL-NEXT:    mov w9, #5826 // =0x16c2
-; GISEL-NEXT:    movk w9, #364, lsl #16
-; GISEL-NEXT:    umull x8, w8, w9
+; GISEL-NEXT:    mov w8, #5826 // =0x16c2
+; GISEL-NEXT:    and w9, w0, #0xff
+; GISEL-NEXT:    movk w8, #364, lsl #16
+; GISEL-NEXT:    umull x8, w9, w8
```
expanding the div generates a bunch of instructions and the order in which they are recombined results in different ouput

tobias-stadler commented on llvm/test/CodeGen/AArch64/GlobalISel/postlegalizer-combiner-sameopcode-hands-crash.mir:
```diff
@@ -9,6 +9,8 @@ body:             |
   bb.1:
     ; CHECK-LABEL: name: crash_fn
     ; CHECK: [[C:%[0-9]+]]:_(p0) = G_CONSTANT i64 0
+    ; CHECK-NEXT: [[C1:%[0-9]+]]:_(s32) = G_CONSTANT i32 0
+    ; CHECK-NEXT: [[C2:%[0-9]+]]:_(s1) = G_CONSTANT i1 false
```
These are dead in the input IR, which doesn't happen inside the CodeGen pipeline.

tobias-stadler commented on llvm/test/CodeGen/AArch64/GlobalISel/postlegalizer-lowering-swap-compare-operands.mir:
```diff
@@ -88,7 +95,8 @@ body:             |
 
     $w0 = COPY %cmp2(s32)
-    RET_ReallyLR implicit $w0
+    $w1 = COPY %cmp1(s32)
```
cmp1 was dead in the input IR, but according to the comment this seems unintended.

tobias-stadler commented on llvm/test/CodeGen/AArch64/setcc_knownbits.ll:
```diff
@@ -33,8 +33,7 @@ define noundef i1 @logger(i32 noundef %logLevel, ptr %ea, ptr %pll) {
 ; CHECK-GI-NEXT:    ldr x8, [x1]
-; CHECK-GI-NEXT:    ldrb w8, [x8]
-; CHECK-GI-NEXT:    and w0, w8, #0x1
+; CHECK-GI-NEXT:    ldrb w0, [x8]
```
The immediate DCEing and better ordering of combines prevents the PreLegalizerCombiner from generating a bunch of useless artifacts that it can't combine away again. These artifacts are converted into G_AND by the ArtifactCombiner, which can't be combined away by the redundant_and combine, because KnownBits can't look through the implicit anyext of the G_LOAD. We are probably missing some combines that convert the load into sext/zext versions.

tschuett commented:
Is this a fundamental issue of the combiner or do we have to revisit the situation when the combiner becomes slowly more powerful?

tobias-stadler commented:
> Is this a fundamental issue of the combiner or do we have to revisit the situation when the combiner becomes slowly more powerful?

This is a fundamental issue of combiner-style algorithms. Fixed-point iteration just burns too much compile-time for no good reason. Both the InstCombiner and the DAGCombiner don't use fixed-point iteration anymore for the same compile-time reasons. The heuristics I have implemented should be mostly on-par with how thoroughly the other combiner implementations currently handle retrying combines. The main difference is that we can rely on the Observer to detect changes for retrying combines and the other combiner implementations need to use WorkList-aware functions for performing changes on the IR. If this approach is good enough for SelectionDAG, I don't see this approach becoming a problem for GlobalISel.

aemerson reviewed (approved):
These are some very nice improvements, thanks for working on this. None of the test output changes look to be exposing problems with this patch, so LGTM.

tobias-stadler committed:
[spr] changes introduced through rebase

Created using spr 1.3.6-wip

[skip ci]

tobias-stadler committed:
Rebase

Created using spr 1.3.6-wip

tobias-stadler commented:
> These are some very nice improvements, thanks for working on this. None of the test output changes look to be exposing problems with this patch, so LGTM.

Thanks a lot for reviewing this whole stack for me!
